### ISSUE TITLE
(CLI) Runner does not report AppDomain unloading timeout

### ISSUE THREAD

**Issue by BrunoJuchli** [nunit/nunit#329]
Currently `DomainManager.DomainUnloader.Unload` is silently* timing `AppDomain.Unload(domain) ` out after 30 seconds.

* an entry like `09:35:36.420 Error [ 1] DomainManager: Unable to unload AppDomain test-domain-FOOBAR.dll, Unload thread timed out` is logged to the NUnit trace log (%AppData%\..\Local\NUnit\logs.

[msdn](http://msdn.microsoft.com/en-us/library/system.appdomain.unload%28v=vs.110%29.aspx) states that

>  If a thread does not abort, for example because it is executing unmanaged code, or because it is executing a finally block, then after a period of time a CannotUnloadAppDomainException is thrown in the thread that originally called Unload. If the thread that could not be aborted eventually ends, the target domain is not unloaded. Thus, in the .NET Framework version 2.0 domain is not guaranteed to unload, because it might not be possible to terminate executing threads. 

To sum it up:
* the `AppDomain` is not necessarily unloaded when `DomainUnloader.Unload` finishes
 * could this lead to issues when running other test assemblies subsequently?
 * it is still unloaded, however, when the process is exited (which is still possible)
* the issue is not made visible to the user.
* There is no statement made by microsoft how long `.Unload(domain)` takes before it times out with an `CannotUnloadAppDomainException`
* `CannotUnloadAppDomainException` may indicate serious issues in either productive or test code
 * i don't know about the part "for example because it is executing unmanaged code"... i think when a unit test ends, so should all processing of the unit under test.

Also see https://groups.google.com/forum/#!topic/nunit-discuss/Y_qy00H44ZQ

**Comment by CharliePoole** [comment]
This has been around for a long time and was low priority. I just raised it because it causes  TFS runs to fail when it occurs. See nunit/nunit3-vs-adapter#177

**Comment by CharliePoole** [comment]
I've done some investigating. The "unload thread" is actually doing two things.

1. Unloading the appdomain
2. Erasing the shadow cache

Questions:

1. Do we need to erase the shadow cache? We started doing it years ago because NUnit used to always use it and the cache would grow out of control. Also, we created our own cache location in order to make it easier to erase. We no longer default to using the shadow cache and I suspect very few people use the option that enables it. Therefore, could we not forget entirely about establishing our own cache location and subsequently erasing it? Seems like the OS should take care of this.

2. Do we want to have visibility when the AppDomain fails to load (because some user thread is still running in it) or do we prefer to silently ignore the problem.

If we answer these questions, it may be pretty easy to put in a fix to this very old issue.

**Comment by rprouse** [comment]
For the shadow cache, :+1: 

For the AppDomain, do you mean unload? If so, I am not sure. @BrunoJuchli states the error is trivial and I would agree in this situation, but can anyone think of errors that should cause test failures or should at least be exposed? If not, then :+1:

**Comment by CharliePoole** [comment]
I'll move ahead with the shadow cache. This will simplify things enormously. One problem with the current code is that it removes those files and directories inside a finally block, which is not a good place to be getting an exception.

For the AppDomain, yes, unload - I changed the earlier comment. By the time we are doing this, we have already run the tests and they either passed or failed. The only problem I see is a potential resource issue.

**Comment by BrunoJuchli** [comment]
Failing to unload an appdomain hints at a concurrency issue. Probably the test code has not been synchronized correctly (p.Ex. starting an asynchronous worker, testing that it has done its work but not stopping it). It's entirely possible that while the synchronization is missing all tests pass (for example because one forgot to test it or one chose not to - because it's difficult).

These kinds of issues are some of the hardest to detect and fix.
Therefore it's important that as much information as possible is reported to the user.

Hence i believe that failing to unload the appdomain should fail the test-run (even though it will not fail any specific test).
Since failure to unload the appdomain could conceivably cause further test-runs to fail it may even be an option to "crash" the runner (or whatever there is to report problems of the highest magnitude).

**Comment by CharliePoole** [comment]
`Debugger.Launch` is an interesting idea. It would only affect debug builds and so CI builds would not be broken. Broken CI builds is the main reason people complained about this.

**Comment by BrunoJuchli** [comment]
Though they shouldn't complain in the first place - i'm pretty sure 99% of the time it means a bug in their code, not in NUnit's code.
If the cause were clearly visible and not so complex I bet they wouldn't complain. Too bad we can't make it that easy...
An idea might also be to write a WIKI Page containing a few examples/reasons leading to the error and ideas on how to approach a fix. The error text could contain a "link" to that WIKI page.
